napari/view_layers.py

```diff
@@ -139,39 +145,153 @@ def _make_viewer_then(add_method: str, args, kwargs) -> Viewer:
+    shear : 1-D array or n-D array
+        Either a vector of upper triangular values, or an nD shear matrix with
+        ones along the main diagonal.
+    affine : n-D array or napari.utils.transforms.Affine
+        (N+1, N+1) affine transformation matrix in homogeneous coordinates.
+        The first (N, N) entries correspond to a linear transform and
+        the final column is a length N translation vector and a 1 or a napari
+        `Affine` transform object. Applied as an extra transform on top of the
+        provided scale, rotate, and shear values.
+    opacity : float
+        Opacity of the layer visual, between 0.0 and 1.0.
+    blending : str
+        One of a list of preset blending modes that determines how RGB and
+        alpha values of the layer visual get mixed. Allowed values are
+        {'opaque', 'translucent', and 'additive'}.
+    visible : bool
+        Whether the layer visual is currently being displayed.
+    multiscale : bool
+        Whether the data is a multiscale image or not. Multiscale data is
+        represented by a list of array like image data. If not specified by
+        the user and if the data is a list of arrays that decrease in shape
+        then it will be taken to be multiscale. The first image in the list
+        should be the largest. Please note multiscale rendering is only
+        supported in 2D. In 3D, only the lowest resolution scale is
+        displayed.
+    cache : bool
+        Whether slices of out-of-core datasets should be cached upon retrieval.
+        Currently, this only applies to dask arrays.
+    plane : dict or SlicingPlane
+        Properties defining plane rendering in 3D. Properties are defined in
+        data coordinates. Valid dictionary keys are
+        {'position', 'normal', 'thickness', and 'enabled'}.
+    experimental_clipping_planes : list of dicts, list of ClippingPlane, or ClippingPlaneList
+        Each dict defines a clipping plane in 3D in data coordinates.
+        Valid dictionary keys are {'position', 'normal', and 'enabled'}.
+        Values on the negative side of the normal are discarded if the plane is enabled.
+    viewer: Viewer object, optional, by default None.
+    title : string, optional
+        The title of the viewer window. by default 'napari'.
+    ndisplay : {2, 3}, optional
+        Number of displayed dimensions. by default 2.
+    order : tuple of int, optional
+        Order in which dimensions are displayed where the last two or last
+        three dimensions correspond to row x column or plane x row x column if
+        ndisplay is 2 or 3. by default None
+    axis_labels : list of str, optional
+        Dimension names. by default they are labeled with sequential numbers
+    show : bool, optional
+        Whether to show the viewer after instantiation. by default True.
+
+    Returns
+    -------
+    Tuple: (Viwer, Layer)
+    """
+
+    return _make_viewer_then('add_image', args, kwargs, viewer=viewer)
```

Comment:
Assuming we want to keep the channel_axis= behaviour (I think we decided yes), then I think maybe we want this to *always* return a list of layers, which would contain a single layer if `channel_axis=None`, or one or more if `channel_axis=some_int`. Ref:

https://github.com/napari/napari/issues/4693#issuecomment-1200509018

> where layers is a list (not a LayerList) of any added Layer objects - will be 1 if multichannel=False and will be more if multichannel=True.

Thoughts @tlambert03? ie do we want the return type to be variable, or always a list (though it may be cumbersome in the single channel case)? e.g. is this too clunky?

```
viewer, (layer,) = napari.imshow(array)
```

Maybe it is and we leave it as variable?

Reply:
> Thoughts @tlambert03? ie do we want the return type to be variable, or always a list (though it may be cumbersome in the single channel case)? e.g. is this too clunky?

my vote would definitely be to always return a list (and I'd like to see a type annotation on the return as well ` -> Tuple[Viewer, List[Image]]`).  

but I think that check should be done here, rather than in `_make_viewer_then`. 

```python
viewer, layers = _make_viewer_then('add_image', args, kwargs, viewer=viewer)
layers = layers if isinstance(layers, list) else [layers]
return viewer, layers
```

Reply:
It's funny, overnight I was thinking, nah, let's keep it variable, the whole point of a convenience function is to be convenient, not to be type-safe... 😂 Especially given the motivation for this whole thing from #4693 (existing convenience functions are not quite convenient enough)... How strongly do you feel about it @tlambert03?

Reply:
Eh, 6 out of 10 strong?  

I think it's convenient up until the point where you get a type error because you didn't switch over your return object handling when you changed your channel axis argument.  And then it would have been more convenient not to have to remember to change the rest of your code accordingly.

Yes, a big driver for this is convenience, but it kinda got pushed back to the fore because of the channel axis deprecation, which is largely because of the type problem.

But it's not a blocker for me, so you can make the call :)

Reply:
Ok I'm back 😂 

The type thingy is interesting, because I would probably write a lot of `image, (layer,) = ...`, and then you're in the same box as before. (Though in this case the error happens immediately, rather than when trying to access layer attributes later... That's an advantage.) I would argue that in code (vs interactive usage), we would discourage the use of imshow in favour of instantiating a viewer and using `add_` methods.

I'd love to hear from other core devs and users on this... I'll ping on Zulip.

Reply:
As I said, not a blocker for me. So if you're coming around to a strong opinion for variable return types, that's fine

Reply:
After a hearty discussion on Zulip we have two options, either variable return type or variable *length* return type, `return viewer, *layers`. [Here's the summary message on Zulip](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/imshow.20return.20type/near/296525428). People are going to vote on the options and then we'll settle on one. I'm happy with either.